browser/brave_ads/tabs/ads_tab_helper.cc

```diff
@@ -85,6 +81,36 @@ bool AdsTabHelper::IsVisible() const {
   return is_web_contents_visible_ && is_browser_active_.value_or(false);
 }
 
+void AdsTabHelper::MaybeSetBrowserIsActive() {
+  if (is_browser_active_ && *is_browser_active_) {
+    // Already active.
+    return;
+  }
+
+  is_browser_active_ = true;
+
+  MaybeNotifyBrowserDidBecomeActive();
+
+  // Maybe notify tab change after the browser active state changes because
+  // `OnVisibilityChanged` can be called before `OnBrowserSetLastActive`.
+  MaybeNotifyTabDidChange();
+}
+
+void AdsTabHelper::MaybeSetBrowserIsNoLongerActive() {
+  if (is_browser_active_ && !*is_browser_active_) {
+    // Already inactive.
+    return;
+  }
+
+  is_browser_active_ = false;
+
+  MaybeNotifyBrowserDidResignActive();
+
+  // Maybe notify tab change after the browser active state changes because
+  // `OnVisibilityChanged` can be called before `OnBrowserNoLongerActive`.
+  MaybeNotifyTabDidChange();
+}
```

Comment:
Decouple so that we can use the same code for both desktop and Android. There is a cleaner solution which I will work on to ride the trains as part of a followup fix for https://github.com/brave/brave-browser/issues/38464.